lsm: refactor compaction

matklad opened the pull request:
Preface: 

I've been hacking on this code for a past couple of days, and, while I am not yet bold enough to hit `./zig/zig build check` button, I am at the point where I desperately need feedback!

I _think_ the overall shape of code should be somewhat clear at this point, even if it is missing a detail or two (I see that I've removed `release_table_blocks`, I probably should be calling that _somewhere_).

Right now, I am at the stage where I am pretty convinced in this new approach, but that's probably simply because I've been staring at this code for quite some time and it makes _too_ much sense to me. So curious WDYT here!

---

Refactor compaction to:
- make the code easier to follow,
- make it less unobvious how the compactions are actually schedules (that is, what work can be in-flight at any given point in time),
- make scheduling a bit more aggressive inside a single compaction (e.g. start merge when the first two value blocks are in, separate IO and block resources),
- make it easier to run several compactions at the same time.

Code Changes
============

* Fewer types
* All tricky logic is centralized in `compaction_dispatch`
* Many FIFOs are replaced with fewer RingBuffers
* Flatten Beat and Bar into compaction

CompactionPipeline no longer implements pipelining. Now, it only deals with scheduling compactions themselves. Pipelining is the job of each individual compaction. That is, CompactonPipeline kicks off a particualar compaciton, and recieves the control back only when that compaciton's beat is done.

CompactionPipeline should be rename to CompactionSchedule, or maybe just merged with a forest.

A new type, CompactionPool, is introduced. This is a bag of all resources shared by compactions. Previously, blocks, reads and writes were lumped together. Now, CompactionPool holds separate IOPS for reads, writes, blocks, and CPUS.

CompactionPool is a forest-level object, but it _doesn't_ depend on the type of the forest. So, all compactions share the same CompactionPool _type_, which allows them to share the actual object.

The actual pipelining logic is moved to `Compaction.compaction_dispatch` function. This function no longer tries to _explicitly_ encode a pipeline. Rather, it is just a loop that spawns a bunch of jobs, and the pipelining happens as a side-effect of that. The pseudo-code for compaction_dispatch is:

    for each job:
        if have resources for the job AND
           job's dependencies are fulfilled AND
           it makes sense to start the job:

           start the job

Here, the jobs are reading the index blocks, reading the value blocks, merging/copying, writing the data blocks, writing the index blocks. Resources are blocks (memory) and grid reads&writes. Dependencies are blocks read previously by other jobs (reading a value block needs an index block). The "makes sense" part coves:
- spreading resources across jobs to avoid deadlocks,
- avoiding overreads when finishing a beat-worth of work.

There _isn't_ some kind of comptime meta programming framework for defining jobs, it's just a good old pile of ifs (pushed up to the `compaction_dispatch` cental location).

Dependency tracking is done via RingBuffers. For example, there's a RingBuffer of level_b value blocks. Blocks are added to this queue in order when we _start_ the corrsponding read. When we finish the read, we flip the status of block (which is in the queue at this point) to `read_done`.

So, when we think about launching a merge, we check that:

- there's a head value block in the queue
- the status of this block is `read_done`

If the merge ends up fully consuming the block, it pops it off the queue and adds back to the shared pool.

This setup is similar to a bunch of fifos we used before, but:

* With ring-buffers, so that it's statically obvious what's the upper bound for the size of the queue.
* Queues are matched to block kind (value block/index block/output block) rather than to it's lifetime stage (pending/reading etc).
* There are just fewer queues (I think, didn't actually count).

Behavior Changes
================

* Separate limits for memory, CPU, and IO resources.
* Aggressive scheduling -- start merge as soon as the first two blocks are in
* Running two compactions just work, given enough resources.
* A different "winding down" approach.

Previously, allocating a block for compaction also allocated a grid read and a grid write. In the new approach, those are allocated separately. Additionally, besides available resources there are static limits on the amount of outstanding work. That is, RingBuffers for blocks are statically sized, so you can't read more than X blocks at the same time even if you have spare blocks.

For simplicity, it is assumed that there's actually enough reads and writes to fill the queues fully, but the approach should work even if you restrict _all_ compactions to just a single read and a single write.

The blocks are distributed in a priority round-robin fashion. That is, we first give block to the table builder, and then alternate between level b and level a. This way:

* We avoid deadlocks: if the number of blocks is larger than a minimum, than the jobs will be scheduled in the right order to avoid running out of blocks completely
* We give equal resources to level a and level b (the approach should be extensible to fair allocation, proportional to level "size")
* The schedule ends up being maximally pipelined (the first four reads are: level_a index, level_b index, level_b value, level_a value, which allows starting the merge ASAP).

I _think_ this schedule allows even for a simplistic strategy for running several compactions: if you have N*min blocks, it would be correct to start N compactions at the same time. At least one of them will be guaranteed to be able to proceed. This scheme is simplistic, because:
- a single compaction can monopolize an unfair share of resources. Compactions are greedy.
- there's no code path for "if compaction A frees resource X, wake up compaction B waiting on X".

Both points should be fixable I think though!

Another difference (I think) happens in the "winding down" phase. That is, in what happens when the compaction consumed the required quota of inputs, but haven't yet produced a full value table of output. This is a tricky situation, because, depending on the tombstones, you might have to read quite a bit more tables still. The worst case here is that you have a full table on level A, a full table on level B, but their merge is an empty set.

In the previous approach, due to explicit pipelining, we'll just loop over:

    loop {
        read fixed number of blocks from level
        run merge
    }

In the new approach, this is handled by `AND it makes sense to start the job` part of the pseudocode --- if we _are_ in the winding-down phase, we just artificially limit the depth of our RingBuffers.

Such overreads are tracked via `values_wasted_count`.

matklad commented:
cc @cb22 @sentientwaffle @batiati

matklad commented:
It is getting there!

This now basically works, and passes non-trivial amount of VOPR seeds.

I am yet to restore logging&tracing though...

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -281,253 +277,140 @@ pub fn CompactionType(
-        pub fn init(tree_config: Tree.Config, grid: *Grid, level_b: u8) Compaction {
+        } = .{},
+
+        // Position points at the next vale from the layer that should be feed into the merge
```
```suggestion
        // Position points at the next value from the layer that should be feed into the merge
```

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -65,180 +67,175 @@ pub const compaction_tables_output_max = compaction_tables_input_max;
-                assert(value.stage == .ready);
-                value.stage = .ioing;
-                self.ioing.push(value);
+        pub fn idle(pool: *ResourcePool) bool {
```
When `idle` is true, can we assert that `pool.blocks.count == pool.blocks_backing_storage.len`?
(We should assert it between beats, to make sure that we don't leak blocks.)

matklad replied:
That's what idle used to do, before I "realized" (vopr) that this is wrong! The output index blocks survive between the beats. So `.idle` _only_ assets "IO", and between the bars we separately assert `blocks_acquired() == 0`

sentientwaffle replied:
Ah, good point -- so we can only assert it between bars!

sentientwaffle commented on src/lsm/compaction.zig:
```suggestion
        /// Copy values from values_in to values_out, dropping tombstones as we go.
```

sentientwaffle commented on src/lsm/compaction.zig:
```suggestion
        // plane. To reduce the probability of the optimizer getting confused over pointers, don't
```

sentientwaffle reviewed (commented):
Just a few nits while reading through `compaction.zig`.

matklad committed:
lsm: refactor compaction

Refactor compaction to:
- make the code easier to follow,
- make it less unobvious how the compactions are actually schedules
  (that is, what work can be in-flight at any given point in time),
- make scheduling a bit more aggressive inside a single compaction (e.g.
  start merge when the first two value blocks are in, separate IO and
  block resources),
- make it easier to run several compactions at the same time.

Code Changes
============

* Fewer types
* All tricky logic is centralized in `compaction_dispatch`
* Many FIFOs are replaced with fewer RingBuffers
* Flatten Beat and Bar into compaction

CompactionPipeline no longer implements pipelining. Now, it only deals
with scheduling compactions themselves. Pipelining is the job of each
individual compaction. That is, CompactonPipeline kicks off a
particualar compaciton, and recieves the control back only when that
compaciton's beat is done.

CompactionPipeline should be rename to CompactionSchedule, or maybe just
merged with a forest.

A new type, CompactionPool, is introduced. This is a bag of all resources
shared by compactions. Previously, blocks, reads and writes were lumped
together. Now, CompactionPool holds separate IOPS for reads, writes,
blocks, and CPUS.

CompactionPool is a forest-level object, but it _doesn't_ depend on the
type of the forest. So, all compactions share the same CompactionPool
_type_, which allows them to share the actual object.

The actual pipelining logic is moved to `Compaction.compaction_dispatch`
function. This function no longer tries to _explicitly_ encode a
pipeline. Rather, it is just a loop that spawns a bunch of jobs, and the
pipelining happens as a side-effect of that. The pseudo-code for
compaction_dispatch is:

    for each job:
        if have resources for the job AND
           job's dependencies are fulfilled AND
           it makes sense to start the job:

           start the job

Here, the jobs are reading the index blocks, reading the value blocks,
merging/copying, writing the data blocks, writing the index blocks.
Resources are blocks (memory) and grid reads&writes. Dependencies are
blocks read previously by other jobs (reading a value block needs an
index block). The "makes sense" part coves:
- spreading resources across jobs to avoid deadlocks,
- avoiding overreads when finishing a beat-worth of work.

There _isn't_ some kind of comptime meta programming framework for
defining jobs, it's just a good old pile of ifs (pushed up to the
`compaction_dispatch` cental location).

Dependency tracking is done via RingBuffers. For example, there's a
RingBuffer of level_b value blocks. Blocks are added to this queue in
order when we _start_ the corrsponding read. When we finish the read, we
flip the status of block (which is in the queue at this point) to
`read_done`.

So, when we think about launching a merge, we check that:

- there's a head value block in the queue
- the status of this block is `read_done`

If the merge ends up fully consuming the block, it pops it off the queue
and adds back to the shared pool.

This setup is similar to a bunch of fifos we used before, but:

* With ring-buffers, so that it's statically obvious what's the upper
  bound for the size of the queue.
* Queues are matched to block kind (value block/index block/output block)
  rather than to it's lifetime stage (pending/reading etc).
* There are just fewer queues (I think, didn't actually count).

Behavior Changes
================

* Separate limits for memory, CPU, and IO resources.
* Aggressive scheduling -- start merge as soon as the first two blocks
  are in
* Running two compactions just work, given enough resources.
* A different "winding down" approach.

Previously, allocating a block for compaction also allocated a grid read
and a grid write. In the new approach, those are allocated separately.
Additionally, besides available resources there are static limits on the
amount of outstanding work. That is, RingBuffers for blocks are
statically sized, so you can't read more than X blocks at the same time
even if you have spare blocks.

For simplicity, it is assumed that there's actually enough reads and
writes to fill the queues fully, but the approach should work even if
you restrict _all_ compactions to just a single read and a single write.

The blocks are distributed in a priority round-robin fashion. That is,
we first give block to the table builder, and then alternate between
level b and level a. This way:

* We avoid deadlocks: if the number of blocks is larger than a minimum,
  than the jobs will be scheduled in the right order to avoid running
  out of blocks completely
* We give equal resources to level a and level b (the approach should be
  extensible to fair allocation, proportional to level "size")
* The schedule ends up being maximally pipelined (the first four reads
  are: level_a index, level_b index, level_b value, level_a value,
  which allows starting the merge ASAP).

I _think_ this schedule allows even for a simplistic strategy for
running several compactions: if you have N*min blocks, it would be
correct to start N compactions at the same time. At least one of them
will be guaranteed to be able to proceed. This scheme is simplistic,
because:
- a single compaction can monopolize an unfair share of resources.
  Compactions are greedy.
- there's no code path for "if compaction A frees resource X, wake up
  compaction B waiting on X".

Both points should be fixable I think though!

Another difference (I think) happens in the "winding down" phase. That
is, in what happens when the compaction consumed the required quota of
inputs, but haven't yet produced a full value table of output. This is a
tricky situation, because, depending on the tombstones, you might have
to read quite a bit more tables still. The worst case here is that you
have a full table on level A, a full table on level B, but their merge
is an empty set.

In the previous approach, due to explicit pipelining, we'll just loop
over:

    loop {
        read fixed number of blocks from level
        run merge
    }

In the new approach, this is handled by `AND it makes sense to start the
job` part of the pseudocode --- if we _are_ in the winding-down phase,
we just artificially limit the depth of our RingBuffers.

Such overreads are tracked via `values_wasted_count`.

matklad commented on src/lsm/compaction.zig:
```diff
@@ -281,1843 +302,1607 @@ pub fn CompactionType(
-        pub fn deinit(compaction: *Compaction) void {
-            _ = compaction;
-            // TODO: Assert things here - compaction doesn't own anything that needs to be
-            // deallocated.
```
vopr calls deinit on crash, so there isn't anything useful we can assert here.

matklad commented on src/lsm/compaction.zig:
```diff
@@ -281,1843 +302,1607 @@ pub fn CompactionType(
+                if (compaction.tree.table_immutable.mutability.immutable.flushed) {
+                    assert(compaction.quotas.bar == 0);
+                    assert(compaction.quotas.bar_done());
+                    log.debug("{s}:{}: bar_commence: immutable table flushed", .{
```
I am using `$name:$level: $method` naming scheme for logs

matklad commented on src/trace.zig:
```diff
@@ -106,9 +106,8 @@ const trace_span_size_max = 1024;
     replica_commit,
 
-    compact_blip_read,
-    compact_blip_merge,
```
It's not clear how to trace read and writes --- they aren't really spans, it's more like we have `n` concurrent IOPS in flight at any given moment.

matklad commented on src/vsr/replica_test.zig:
```diff
@@ -59,6 +59,15 @@ comptime {
+    var c = t.clients(0, t.cluster.clients.len);
+    try c.request(checkpoint_2_trigger, checkpoint_2_trigger);
+    try expectEqual(t.replica(.R_).commit(), checkpoint_2_trigger);
+}
```
I personally love to leave such smoke tests in, as they make it easier to debug stuff during larger refactors!

matklad commented on src/vsr/replica.zig:
```diff
@@ -7463,9 +7463,7 @@ pub fn ReplicaType(
-                if (journal_header != null and vsr.Headers.dvc_header_type(header) == .valid and
-                    journal_header.?.checksum != header.checksum)
-                {
+                if (journal_header != null and journal_header.?.checksum != header.checksum) {
```
That's a revert of #2385 which fixes correctness issue at the cost of adding a liveness one. Will back out before merging.

chaitanyabhandari replied:
And the liveness issue is being addressed in https://github.com/tigerbeetle/tigerbeetle/pull/2399!

matklad commented:
Was pretty green before the last round of cleanups, 

<img width="1309" alt="Screenshot 2024-10-08 at 21 30 08" src="https://github.com/user-attachments/assets/aefd77d7-11bb-4776-8eb0-4788621a101b">

the benchmark results are not terrible (though I haven't properly benchmarked this yet), and I _think_ I did everything I wanted to in the scope of this PR!

Ready for review!

sentientwaffle commented on src/lsm/forest.zig:
```diff
@@ -756,635 +738,169 @@ pub fn ForestType(comptime _Storage: type, comptime groove_cfg: anytype) type {
+        const ResourcePool = ResourcePoolType(Grid);
+
+        // The minimum number of blocks required to complete all compactions. This is the sum of
+        // blocks reserved between the bars plus enough blocks to complete a single bar.
```
This comment doesn't refer to `forest_compactions_concurrent_max`. (I think it is referring to `forest_block_count_min`.)

sentientwaffle commented on src/lsm/forest.zig:
```diff
@@ -756,635 +738,169 @@ pub fn ForestType(comptime _Storage: type, comptime groove_cfg: anytype) type {
 
-                // Iterate by levels first, then trees, as we expect similar levels to have similar
-                // time-of-death for writes. This helps internal SSD GC.
+            if (first_beat or half_beat) {
```
Can we assert `pool.blocks_acquired() == 0` here?

matklad committed:
lsm: spice up tree fuzzer

Exercise compaction more thoroughly through:

- allocating a variable number of blocks for the bool,
- actually pacing it over several beats,
- varying the pacing abruptly

sentientwaffle commented on src/lsm/forest.zig:
```diff
@@ -428,16 +427,7 @@ pub fn ForestType(comptime _Storage: type, comptime groove_cfg: anytype) type {
-                assert(forest.compaction_pipeline.compactions.count() == 0);
-                assert(forest.compaction_pipeline.bar_active.count() == 0);
-            }
+            forest.compaction_schedule.beat_start(op, compact_trees_callback);
```
Nit, but `manifest_log.compact()` takes `(callback, op)`, so maybe swap the order here to match?

matklad replied:
We've decided that callbacks go last:

>- Callbacks go last in the list of parameters. This mirrors control flow: callbacks are also
  _invoked_ last.

I will actually swap it here for consistency, and then send a follow up to correct all the callbacks I could find, as per our zero tech-debt rule which we are violating here (had we fixed callbacks before, we wouldn't have to exchange the comments here (TigerStyle is _so_ good, we should try to follow it better :-) )).

sentientwaffle replied:
> go last in the list of parameters

Oops; I thought this was the other way around in tigerstyle!

sentientwaffle commented on src/lsm/forest.zig:
```diff
@@ -756,635 +738,169 @@ pub fn ForestType(comptime _Storage: type, comptime groove_cfg: anytype) type {
+        const ResourcePool = ResourcePoolType(Grid);
+
+        const forest_compactions_concurrent_max =
+            stdx.div_ceil(constants.lsm_levels, 2) * Forest.tree_infos.len;
+        const forest_block_count_bar_max =
+            forest_compactions_concurrent_max * compaction_block_count_bar_max;
+        /// The minimum number of blocks required to complete all compactions. This is the sum of
+        /// blocks reserved between the bars plus enough blocks to complete a single beat.
+        const forest_block_count_min = forest_block_count_bar_max +
+            // beat_min includes bar_max for that beat.
+            (compaction_block_count_beat_min - compaction_block_count_bar_max);
+
         pub fn init(
-            self: *CompactionPipeline,
+            self: *CompactionSchedule,
             allocator: mem.Allocator,
             grid: *Grid,
             forest: *Forest,
             block_count: u32,
         ) !void {
```
Out of scope, but right now VOPR randomizes the number of compaction blocks, but uses the same value for all members of the cluster:

```
            .accounting => .{
                .batch_size_limit = batch_size_limit,
                .lsm_forest_compaction_block_count = random.uintAtMost(u32, 256) +
                    StateMachine.Forest.Options.compaction_block_count_min,
                .lsm_forest_node_count = 4096,
                .cache_entries_accounts = 256,
                .cache_entries_transfers = 256,
                .cache_entries_posted = 256,
                .cache_entries_account_balances = 256,
            },
```

If we instead randomized that per-replica, the storage checker would verify that the output of compaction is independent of the compaction block count.

sentientwaffle commented on src/lsm/forest.zig:
```diff
@@ -756,635 +738,169 @@ pub fn ForestType(comptime _Storage: type, comptime groove_cfg: anytype) type {
-            for (self.compactions.slice(), 0..) |*compaction, i| {
-                if (!self.bar_active.isSet(i)) continue;
-                if (compaction.move_table) continue;
+        fn beat_resume(self: *CompactionSchedule) void {
```
```suggestion
        fn beat_resume(self: *CompactionSchedule) void {
            assert(self.callback != null);

```

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -63,182 +65,202 @@ pub const compaction_tables_input_max = 1 + constants.lsm_growth_factor;
+            },
+            // The index block is freed immediately after the read for the last value block is
+            // submitted so, by the time merge consumes a value block, the index block is gone
+            // already. Use this field to mark the spots when the compaciton position should advance
```
```suggestion
            // already. Use this field to mark the spots when the compaction position should advance
```

... But "compaciton" is a perfect/terrible name for the smallest possible unit of compaction. :laughing:

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -281,1843 +302,1617 @@ pub fn CompactionType(
+            assert(compaction.idle());
+            assert(compaction.stage == .inactive);
+            compaction.stage = .paused;
+            compaction.op_min = compaction_op_min(op);
```
Since this is being called only at the beginning of each (half-)bar, I think we can assert:
```
assert(op == compaction_op_min(op));
```

matklad replied:
Ah, this op/op_min stuff did look odd to me, but I haven't looked deeper, you are of course correct! 

We probably should _also_ update the diagram above `compaction_op_min`, and the rest of the docs, but I've postponed docs refreshing for now.

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -281,1843 +302,1617 @@ pub fn CompactionType(
+        // as the read for the last value block is scheduled, the pipeline should not dry out even
+        // when switching between the tables.
+        //
+        // Note that level_L_position field tracks the logical, deterministic progression of
```
```suggestion
        // Note that level_L_position fields track the logical, deterministic progression of
```

Or maybe

```suggestion
        // Note that level_{a,b}_position fields track the logical, deterministic progression of
```

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -281,1843 +302,1617 @@ pub fn CompactionType(
-                        range_b,
-                    ),
+            // Append the entries to the manifest update queue here and now if we're doing
+            // move table. They'll be applied later by bar_apply_to_manifest.
```
```suggestion
            // move table. They'll be applied later by bar_complete().
```

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -281,1843 +302,1617 @@ pub fn CompactionType(
-                    compaction.bar.?.target_values_write_count =
-                        compaction.bar.?.compaction_tables_value_count;
+            if (compaction.table_info_a == null) {
+                assert(compaction.range_b == null);
```
```suggestion
                assert(compaction.range_b == null);
                assert(compaction.manifest_entries.count == 0);
```

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -281,1843 +302,1617 @@ pub fn CompactionType(
-            assert(beats_max > 0);
+            // Mark the immutable table as flushed, if we were compacting into level 0.
+            if (compaction.level_b == 0) {
+                compaction.tree.table_immutable.mutability.immutable.flushed = true;
```
```suggestion
                assert(!compaction.tree.table_immutable.mutability.immutable.flushed);
                compaction.tree.table_immutable.mutability.immutable.flushed = true;
```

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -281,1843 +302,1617 @@ pub fn CompactionType(
+                    compaction.level_b,
+                });
+                return;
+            }
```
```suggestion
            }
            assert(!compaction.move_table);
```

sentientwaffle commented on src/lsm/compaction.zig:
So since the safety counter has a `+1`, if the safety counter is exhausted this `!progressed` assert is guaranteed to succeed?

matklad replied:
Ah, good catch, I am missing `else unreachable` here! The `assert(!progressed)` here is mostly just for readability, it is an `endwhile` of sorts that reminds the reader, by the time they get to this line, that there's this whole `for` loop thing going on, which started 200 lines before.

I am also torn on what we should do if the safety counter is exhausted: crash with `unreachable`, or `log.error` and soldier on? In this case, the second option would be correct. But I figured that crashing is probably better as a principle.

matklad committed:
lsm: address review comments

Co-authored-by: djg <sentientwaffle@gmail.com>

sentientwaffle commented on src/lsm/compaction.zig:
Ah, so we use `free_set.release()` instead of `grid.release()` b/c we don't want to demote the blocks from the cache, as they may be re-read again during the next beat(s).

Let's assert in `bar_commence()` for each (level A and B) index block address: 

```
assert(!compaction.grid.free_set.is_released(address));
```

matklad replied:
>Ah, so we use free_set.release() instead of grid.release() b/c we don't want to demote the blocks from the cache, as they may be re-read again during the next beat(s).

No, that is a genuine refactoring bug! I didn't intend to fix this.

That being said, this feels a bit backwards --- it seems that, rather than demoting tables after the fact, the grid should perhaps read them without bumping the cache in the first place? 

>Let's assert in bar_commence() for each (level A and B) index block address:

Ah, this is great!

sentientwaffle commented on src/lsm/compaction.zig:
`blip_merge` is gone:

```suggestion
                // Only consume one block at a time so that a beat never outputs past its quota
                // by more than 1 value block.
```

sentientwaffle commented on src/lsm/compaction.zig:
FIXME :smile: 
(line 1422 also)

matklad replied:
oh dear, I guess one more tidy check is coming soon!

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -281,1843 +302,1660 @@ pub fn CompactionType(
-                    i += 1;
+                // Read level A value block.
+                if (compaction.table_info_a.? == .immutable) {
+                    // The whole table is in memory, no need to read anything.
```
```suggestion
                    // The whole table is in memory, no need to read anything.
                    assert(compaction.level_a_index_block.count == 0);
```

(I found this assert useful when considering the "priority" calculations for reading value blocks.)

matklad commented:
Rebased to pick up the VOPR fix for "all clients evicted" issue

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -281,1843 +302,1648 @@ pub fn CompactionType(
-        pub fn blip_read(compaction: *Compaction, callback: BlipCallback, ptr: *anyopaque) void {
-            log.debug("blip_read({s}): scheduling read IO", .{compaction.tree_config.name});
+            // The loop below runs while (!progressed) and, every time progressed is set to true,
+            // one of the safety_counter resources is acquired.
```
Isn't it actually

```suggestion
            // The loop below runs while (progressed) and, every time progressed is set to true,
            // one of the safety_counter resources is acquired.
```

Or it could be

```suggestion
            // The loop below runs until (!progressed) and, every time progressed is set to true,
            // one of the safety_counter resources is acquired.
```

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -281,1843 +302,1648 @@ pub fn CompactionType(
+            const safety_counter =
+                compaction.pool.?.reads.total() +
+                compaction.pool.?.writes.total() +
+                compaction.pool.?.cpus.total() + 1;
```
Or maybe even
```suggestion
                compaction.pool.?.reads.available() +
                compaction.pool.?.writes.available() +
                compaction.pool.?.cpus.available() + 1;
```

matklad replied:
Hm, I was mentally trying to convince you that `total` is better, but ended up convincing myself that `available` is the way to go 🤣 

I can't put my finger on the exact guideline here, it seems the two things are in tension here:

* we want to be our safety counters to be simple, so it is obvious that they are correct.
* but we also want them to be tight, such that, if they are _not_ correct, the VOPR tells us this. 

Here, total is less tight, but simpler, but I think it's not _that_ simpler than `.available`, and that better captures the reason _why_ this is valid safety counter.

sentientwaffle commented on src/lsm/compaction.zig:
```suggestion
        fn merge_inputs(compaction: *const Compaction) struct { ?[]const Value, ?[]const Value } {
```

matklad replied:
Fixed&squashed all review commits down to one!

sentientwaffle reviewed (dismissed):
LGTM!
